File: agent/x402-demo/src/registry.ts
/**
 * Minimal AgentRegistry client — register an agent and (as admin) record payments
 * so reputation scores bump when a bounty is released. Signs headlessly with a
 * keypair; the contract spec is fetched from chain (no generated bindings needed).
 */

import { Keypair, contract } from "@stellar/stellar-sdk";
import { basicNodeSigner } from "@stellar/stellar-sdk/contract";

const RPC_URL = process.env.SOROBAN_RPC ?? "https://soroban-testnet.stellar.org";
const NETWORK_PASSPHRASE =
  process.env.NETWORK_PASSPHRASE ?? "Test SDF Network ; September 2015";

async function registryClient(secret: string) {
  const kp = Keypair.fromSecret(secret);
  const signer = basicNodeSigner(kp, NETWORK_PASSPHRASE);
  const contractId = process.env.AGENT_REGISTRY_ID;
  if (!contractId) throw new Error("AGENT_REGISTRY_ID not set");
  const client = await contract.Client.from({
    contractId,
    rpcUrl: RPC_URL,
    networkPassphrase: NETWORK_PASSPHRASE,
    publicKey: kp.publicKey(),
    signTransaction: signer.signTransaction,
  });
  return { client: client as unknown as Record<string, (...a: unknown[]) => Promise<any>>, address: kp.publicKey() };
}

/** Register the agent (signs with its own key). Idempotent: no-ops if already registered. */
export async function registerAgent(
  secret: string,
  name: string,
  endpoint: string,
  description: string
): Promise<{ address: string; alreadyRegistered: boolean }> {
  const { client, address } = await registryClient(secret);
  const existing = await client.get_agent({ agent: address });
  if (existing.result) return { address, alreadyRegistered: true };
  const tx = await client.register({ agent: address, name, endpoint, description });
  await tx.signAndSend();
  return { address, alreadyRegistered: false };
}

/** Read an agent profile (or null). Uses `viewerSecret` only as the read source. */
export async function getAgent(agent: string, viewerSecret: string) {
  const { client } = await registryClient(viewerSecret);
  const tx = await client.get_agent({ agent });
  const p = tx.result;
  if (!p) return null;
  return {
    address: p.address,
    name: p.name,
    score: String(p.score),
    bountiesDone: Number(p.bounties_done),
  };
}

/** Admin-only: record a payment so the agent's score increases. `adminSecret` must be the registry admin. */
export async function recordPayment(adminSecret: string, agent: string, amount: bigint | string) {
  const { client, address } = await registryClient(adminSecret);
  const tx = await client.record_payment({ caller: address, agent, amount: BigInt(amount) });
  const sent = await tx.signAndSend();
  return { txHash: sent.sendTransactionResponse?.hash };
}

export async function getLeaderboard(viewerSecret: string, limit = 10) {
  const { client } = await registryClient(viewerSecret);
  const tx = await client.get_leaderboard({ limit });
  return ((tx.result ?? []) as Array<[string, bigint]>).map(([address, score]) => ({
    address,
    score: String(score),
  }));
}

File: agent/mcp-server/registry.py
"""
QuestBoard agent registry helpers — AgentRegistry contract interactions.
"""

from stellar_sdk import Server, scval

TESTNET_RPC = "https://soroban-testnet.stellar.org"
TESTNET_PASSPHRASE = "Test SDF Network ; September 2015"


def get_leaderboard(contract_id: str, limit: int = 10) -> list:
    """Call AgentRegistry.get_leaderboard()."""
    if not contract_id:
        return []

    server = Server(TESTNET_RPC)
    contract = scval.to_address(contract_id)

    try:
        # Placeholder: real impl simulates get_leaderboard and parses
        # Vec<(Address, i128)>
        return []
    except Exception as e:
        return [{"error": str(e)}]


def get_agent_profile(contract_id: str, agent: str) -> dict:
    if not contract_id:
        return {}

    server = Server(TESTNET_RPC)
    contract = scval.to_address(contract_id)

    try:
        # Placeholder: real impl simulates get_agent and parses Option<AgentProfile>
        return {}
    except Exception as e:
        return {"error": str(e)}


def register_agent(contract_id: str, agent: str, secret: str, name: str, endpoint: str, description: str) -> dict:
    """Build + submit an AgentRegistry.register transaction."""
    from stellar_sdk import Keypair, TransactionBuilder

    server = Server(TESTNET_RPC)
    keypair = Keypair.from_secret(secret)
    account = server.load_account(agent)
    contract = scval.to_address(contract_id)

    tx = (
        TransactionBuilder(account, network_passphrase=TESTNET_PASSPHRASE, base_fee=100_000)
        .append_invoke_contract_function_op(
            contract=contract,
            function_name="register",
            parameters=[
                scval.to_address(agent),
                scval.to_bytes(name.encode()),
                scval.to_bytes(endpoint.encode()),
                scval.to_bytes(description.encode()),
            ],
        )
        .set_timeout(30)
        .build()
    )
    tx.sign(keypair)
    result = server.submit_transaction(tx)
    if result.get("status") == "ERROR":
        return {"error": str(result)}
    return {
        "tx_hash": result["hash"],
        "status": "registered",
        "agent": agent,
    }